<a href="https://colab.research.google.com/github/dickyadem/MSIB-2-Orbit-Foundation-of-AI/blob/main/K_nearest_neighbors_Enigma_DickyAdeM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Dicky Ade Mahendra <br>
Kelas : Enigma <br>
Perguran Tinggi :  Universitas Gundarma

# K-nearest neighbors 

In [ ]:
# import libraries
import pandas as pd
import itertools
import numpy as np
import matplotlib.pyplot as plt
import numpy as np
from sklearn import preprocessing
from scipy import stats
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

%matplotlib inline

In [ ]:
df = pd.read_csv('http://buku.dioskurn.com/buku1/ch9/churnprediction_ch9.csv', index_col=['customer_id'])
df

product  reload_1  reload_2    video  \
customer_id                                                              
285fae8412c4720a0c79d95f98434422  Kartu A  27734.30  24381.32  22000.0   
f45bce87ca6bf100f222fcc0db06b624  Kartu A  26433.00  26515.50      0.0   
09b54557b1e2a10d998e3473a9ccd2a0  Kartu A  93212.17  67101.83      0.0   
11f252f48be36f93dd429f2ec86cb2f5  Kartu A    183.33   1087.17      0.0   
a8df29ae9195eea348d2f74c967b978d  Kartu A  95296.67  76246.50      0.0   
...                                   ...       ...       ...      ...   
9e8b318d96caa9c0c4a50e8e59f5026c  Kartu B   1634.33  12085.33      0.0   
a310627191bdbded3905ac73e77fe319  Kartu B  30000.33  45170.67      0.0   
b6f11059e5c1df69b8c16d5c39af23dc  Kartu B   3333.33  13338.67      0.0   
88709f1defd232243f729912be696f87  Kartu B  25000.00  33333.33      0.0   
2c5bc32bc9a9c393d393bfc11c409b0d  Kartu C      0.00  11084.00      0.0   

                                    music     games    chat_1    chat_2  \
customer_id                                                               
285fae8412c4720a0c79d95f98434422  33009.9  25669.97    1716.0    2145.0   
f45bce87ca6bf100f222fcc0db06b624      0.0      0.00       0.0   15444.0   
09b54557b1e2a10d998e3473a9ccd2a0      0.0      0.00   86795.5   94649.5   
11f252f48be36f93dd429f2ec86cb2f5      0.0      0.00       0.0       0.0   
a8df29ae9195eea348d2f74c967b978d      0.0  11000.00  118800.0  104940.0   
...                                   ...       ...       ...       ...   
9e8b318d96caa9c0c4a50e8e59f5026c      0.0      0.00       0.0       0.0   
a310627191bdbded3905ac73e77fe319      0.0      0.00       0.0   20001.0   
b6f11059e5c1df69b8c16d5c39af23dc      0.0    872.00       0.0       0.0   
88709f1defd232243f729912be696f87      0.0  23497.33       0.0       0.0   
2c5bc32bc9a9c393d393bfc11c409b0d      0.0      0.00       0.0       0.0   

                                  socmed_1  socmed_2  internet  days_active  \
customer_id                                                                   
285fae8412c4720a0c79d95f98434422       0.0     792.0   11000.0           15   
f45bce87ca6bf100f222fcc0db06b624       0.0       0.0   74151.0           13   
09b54557b1e2a10d998e3473a9ccd2a0     330.0    1485.0   27467.0           15   
11f252f48be36f93dd429f2ec86cb2f5       0.0       0.0       0.0            2   
a8df29ae9195eea348d2f74c967b978d       0.0       0.0   63855.0           15   
...                                    ...       ...       ...          ...   
9e8b318d96caa9c0c4a50e8e59f5026c       0.0       0.0       0.0            1   
a310627191bdbded3905ac73e77fe319       0.0       0.0       0.0            1   
b6f11059e5c1df69b8c16d5c39af23dc       0.0       0.0       0.0            1   
88709f1defd232243f729912be696f87       0.0       0.0       0.0           15   
2c5bc32bc9a9c393d393bfc11c409b0d       0.0       0.0       0.0            1   

                                  tenure  churn  
customer_id                                      
285fae8412c4720a0c79d95f98434422     776      0  
f45bce87ca6bf100f222fcc0db06b624     352      0  
09b54557b1e2a10d998e3473a9ccd2a0    1987      0  
11f252f48be36f93dd429f2ec86cb2f5     285      0  
a8df29ae9195eea348d2f74c967b978d    1081      0  
...                                  ...    ...  
9e8b318d96caa9c0c4a50e8e59f5026c     490      1  
a310627191bdbded3905ac73e77fe319    3120      1  
b6f11059e5c1df69b8c16d5c39af23dc     483      1  
88709f1defd232243f729912be696f87     786      1  
2c5bc32bc9a9c393d393bfc11c409b0d     222      1  

[10000 rows x 14 columns]

In [ ]:
df['product'] = df['product'].map({'Kartu A': 0,'Kartu B': 1, 'Kartu C':2})
df['product'].value_counts()
df.keys()

Index(['product', 'reload_1', 'reload_2', 'video', 'music', 'games', 'chat_1',
       'chat_2', 'socmed_1', 'socmed_2', 'internet', 'days_active', 'tenure',
       'churn'],
      dtype='object')

In [ ]:
# atribut predictor
X = df[['product', 'reload_1', 'reload_2', 'video', 'music', 'games', 'chat_1',
       'chat_2', 'socmed_1', 'socmed_2', 'internet', 'days_active', 'tenure']].values

Y = df['churn'].values


In [ ]:
X = preprocessing.StandardScaler().fit(X).transform(X.astype(int))

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size=0.2, random_state=4)
print ('Train set:', X_train.shape,  Y_train.shape)
print ('Test set:', X_test.shape,  Y_test.shape)

Train set: (8000, 13) (8000,)
Test set: (2000, 13) (2000,)


In [ ]:
#MODELLING
k = 4
#Train Model and Predict  
knn = KNeighborsClassifier(n_neighbors = k).fit(X_train,Y_train)
knn

KNeighborsClassifier(n_neighbors=4)

In [ ]:
yhat = knn.predict(X_test)
print(yhat)

[0 1 0 ... 1 0 0]


In [ ]:
#AKURASI
from sklearn import metrics
print("Train set Accuracy Data Training: ", metrics.accuracy_score(Y_train, knn.predict(X_train))) 
print("Test set Accuracy Data Test: ", metrics.accuracy_score(Y_test, yhat)) 

Train set Accuracy Data Training:  0.942375
Test set Accuracy Data Test:  0.9265


Akurasi yang didapatkan dengan k = 4 adalah 0.9265

**DECISION TREE**

In [ ]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier # Import Decision Tree Classifier
from sklearn.model_selection import train_test_split # Import train_test_split function
from sklearn import metrics #Import scikit-learn metrics module for accuracy calculation

In [ ]:
db = pd.read_csv('http://buku.dioskurn.com/buku1/ch9/churnprediction_ch9.csv', index_col=['customer_id'])

# Pada product, nilai atribut Kartu A didefinisikan sebagai 0, Kartu B didefinisikan sebagai 1, dan Kartu C didefinisikan sebagai 2
db['product'] = df['product'].map({'Kartu A': 0,'Kartu B': 1, 'Kartu C':2})
db['product'].value_counts()
db.keys()

Index(['product', 'reload_1', 'reload_2', 'video', 'music', 'games', 'chat_1',
       'chat_2', 'socmed_1', 'socmed_2', 'internet', 'days_active', 'tenure',
       'churn'],
      dtype='object')

In [ ]:
#split dataset in features and target variable

x = db.iloc[:, :13] # Features
y = db.iloc[:, 13] # Target variable

x = preprocessing.StandardScaler().fit(X).transform(X.astype(float))

In [ ]:
# Split dataset into training set and test set
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42) # 80% training and 20% test


print ('Train set:', x_train.shape,  y_train.shape)
print ('Test set:', x_test.shape,  y_test.shape)

Train set: (8000, 13) (8000,)
Test set: (2000, 13) (2000,)


In [ ]:
clf = DecisionTreeClassifier()

# Train Decision Tree Classifer
clf = clf.fit(x_train,y_train)

#Predict the response for test dataset
y_pred = clf.predict(x_test)

In [ ]:
# Model Accuracy, how often is the classifier correct?
from sklearn import metrics
print("Train set Accuracy Data Training: ", metrics.accuracy_score(y_train, clf.predict(x_train))) 
print("Test set Accuracy Data Test: ", metrics.accuracy_score(y_test, y_pred))

Train set Accuracy Data Training:  0.994
Test set Accuracy Data Test:  0.935


Akurasi yang didapat melalui decision tree adalah 0.935, 
Jadi jika dibandingkan dengan Akurasi yang didapatkan dengan k = 4 adalah 0.9265 maka Decision Tree memperupakan akurasi yang tertinggi

In [ ]:
temp =[1,4,5,0,2,6,9,0,8,3,0]
x,i =0,0

while(temp[i] !=0):
  x +=temp[i]
  i += 2
print(x)

25
